### ISSUE TITLE
iOS 6.0.1

### ISSUE THREAD

**Issue by 4nd413** [H2CO3/libipodimport#4]
Hi, when i try to use the code into my app the song was import into iPod library twice.
One correctly and other with zero byte

I have try also addtoipod in ssh but i have the same issue

tnk's 4nd413

**Comment by H2CO3** [comment]
Hello. OS version, device?

**Comment by 4nd413** [comment]
Hi, iOS 6.0.1 device iPhone 4

**Comment by rootd** [comment]
same happens to me. iOS 6.0.1 on iPhone 4

**Comment by newenglander** [comment]
I can confirm this too (6.1, iPod touch 4g). Also the Cydia app Bridge has been updated with the message "Fixed duplicate song import bug on iOS 6."
The duplicate song doesn't play, however, and is greyed out (as OP mentioned, is an empty file). It can be deleted by swiping.
Does anyone have any ideas as to why, and how to avoid this? Something new in the StoreSession framework?
I thought removePlaceholderDownload (not new) might be promising, but didn't have any effect for me.

**Comment by H2CO3** [comment]
Good question. I'll look into this soon as it seems to bug a lot of users. In the meantime, however, feel free to investigate what is going on, perhaps you can find some valuable information too.

**Comment by intarstudents** [comment]
Same here, iPhone 4S running iOS 6.1.

**Comment by newenglander** [comment]
@H2CO3: I've already tried some things. Did a diff of SSDownloadQueue.h in iOS 5 and 6 and saw one method was added: _handleDownloadsRemoved, but didn't get anywhere with that.
Is the best way to go about doing this just to look and the dumped headers and try things out, or is there a more professional way?

**Comment by H2CO3** [comment]
A "more professional way" would be properly reverse engineering the StoreKit framework (et al.) but in my opinion, the "while (!works) { try out other stuff }" approach is generally faster.

**Comment by newenglander** [comment]
Is it possible to hook into private framework functions with a tweak and do a stack trace, or would most information be missing in release versions?

**Comment by H2CO3** [comment]
That depends on various conditions, I don't know it off the top of my head, honestly.

**Comment by newenglander** [comment]
I tried a bunch of different functions out today, including an alternative to importing with SSDownloadQueue: SSDownloadManager. All had no effect on the double import. Also tried this out on my newly purchased (used) iPhone 4 running iOS 6.1 (normally used iPod touch 4) and for some reason the code runs without creating duplicates in the iPod library, or at least duplicates disappears from view quite quickly. But the same code runs with duplicates on the iPod touch.

Had an alternative workaround: how about deleting the empty file from the iTunes library file (MediaLibrary.sqlitedb)? I had a quick look at the database structure and it would seem quite easy to just delete the empty duplicate.

**Comment by H2CO3** [comment]
The problem is that you can't "just delete" files. If I could, I would have done that already. But it's secured at multiple stages: checksums, whatever. And only the StoreServices framework knows how to do that.

**Comment by intarstudents** [comment]
That sucks. Well the dirty option is to just remove zero sized file entries from database https://gist.github.com/intarstudents/4725159 - thats works somehow

**Comment by H2CO3** [comment]
No you **cannot.** The database is cheksummed, that's what I'm trying to say. If you alter the database, iPod will erase it upon first launch.

**Comment by newenglander** [comment]
I think @H2CO3's right, I tried out @intarstudents' script and it even though it worked at first sight--the empty file was back after a restart, so the database must have been replaced with a backup after failing the checksumming. Or did you have a different experience, @intarstudents?

**Comment by intarstudents** [comment]
@newenglander no, I forgot to test, what will happen after restart, so my bad.

**Comment by newenglander** [comment]
I got a workaround that looks good, using the `ML3MusicLibrary` class from the `MusicLibrary` private framework. There you can open the database with `initWithPath:enableWrites:` and then with `executeSQL:` I ran @intarstudents statements (after waiting a few seconds, because the empty files don't appear immediately) and the empty files remain deleted even after a device restart.

If you guys are interested in trying this out, I could post some code in my yet unchanged fork. But there's not much to it.

Also tried out these functions from the same class, but didn't have any luck with them:

``` removeOrphanedTracks
deleteAutoFilledTracksOfAtLeastTotalSize:
removeLocationsForItemsMissingAssets
resetAllContents
updateTrackIntegrity
resetUbiquitousMetadata
```

**Comment by intarstudents** [comment]
Sounds interesting. I'm interested. Maybe you could provide compiled binaries too?

**Comment by newenglander** [comment]
The problem with providing a binary is I'm using the code for another project. If you're curious, check out my blog: 
http://erike.net/audioexplorer/
If you'd like to try a beta version of that, let me know.

**Comment by newenglander** [comment]
@H2CO3 is strangely silent, not satisfied with the solution? ;) I would just add more conditions to @intarstudents' SQL statements to really make sure we're not deleting anything that was there before (who knows, maybe there are files there with file size of 0).

**Comment by H2CO3** [comment]
Well, I was just busy, I'll try this (but yeah, this seems quite a bit of a hack anyway).

**Comment by newenglander** [comment]
yeah, def. a temporary solution, but still a solution....

**Comment by H2CO3** [comment]
Yup. Lemme see what it does :)

**Comment by H2CO3** [comment]
This just crashed SpringBoard.

**Comment by newenglander** [comment]
Hmm, didn't have any crashing. But I'm using your suggestion about using the entitlements for accessing the iTunes library, so I don't have to hook into SpringBoard. Don't know if that's the issue.

**Comment by eni9889** [comment]
It seems gremlin doesn't have this issue even though they use the same import mechanism

**Comment by H2CO3** [comment]
More accurately, Yussef uses a slightly different piece of code.

**Comment by eni9889** [comment]
True I looked through his code but I couldn't find anything that stood out. Even tried adapting libipodimport to run as a daemon. I do not know if the code on his github is the same as the final build on bigboss though. Would be nice to figure this out maybe it would be better to make libipodimport a daemon service as outlined by apple docs that way it also removes the need for MobileSubstrate

**Comment by H2CO3** [comment]
I'd rather not make a daemon - it functions almost the same as a MobileSubstrate-injected dylib, but then I need to worry about signing with entitlements and setting up the startup plist. It's superfluous. However, I _did_ notice some very small differences in Gremlin which I'll try to implement accordingly to see if it improves anything.

**Comment by eni9889** [comment]
Makes sense. Hopefully it works. There is more completion work in gremlin. I did notice that itunes was throwing some warnings saying that art rendering wasn't done with libipodimport but not with gremlin so maybe its the cleanup thats doing this

**Comment by eni9889** [comment]
If you class-dump the latest itunes import plugin:

@interface GRiTunesImportHelper 
- (BOOL)importAudioFileAtPath:(id)fp8 mediaKind:(id)fp12 withMetadata:(id)fp16;
- (void)_waitForDownloadCompletion:(id)fp8;
- (id)_downloadWithDownloadMetadata:(id)fp8 path:(id)fp12;
- (id)_downloadWithDownloadMetadataLegacy:(id)fp8 path:(id)fp12;
- (id)_downloadMetadataForItemAtPath:(id)fp8 mediaKind:(id)fp12 withMetadata:(id)fp16;
- (id)downloadQueue;

@end

and 

#import <GRiTunesImporter/HTTPConnection.h>

@interface GRSSConnection : HTTPConnection
{
    BOOL _connected;
}
- (void)sendResponseHeadersAndBody;

@end

I think its possible that one has to fake downloading the tone from a server as opposed to just downloading it from disk. I think whats happening is every time its imported that original file gets delete and then it can't find it again thats why it has an error but that's just a guess

**Comment by eni9889** [comment]
That did it :) Song only got added once when I gave it a remote url and I got it to sync back to iTunes :) I can contribute the changes to your repo if ud like. I am having trouble building it with your setup do you mind if I move it to either theos or iosopendev?

**Comment by H2CO3** [comment]
I do mind. However, you don't have to class-dump nor reverse engineer anything :) https://github.com/yfrancis/gremlin/blob/master/plugins/itunes/GRiTunesImportHelper.m

**Comment by eni9889** [comment]
But he doesn't use a server there does he? If you do a class-dump of the itunes plugin now on the phone with the latest gremlin you will see that he is now using CocoaHTTPServer  (HTTPConnection and HTTPServer)  and I just ran my own server and it all works.

**Comment by eni9889** [comment]
I logged the process of SSDownload maybe this will help:

https://gist.github.com/eni9889/941d86201861a856a7fd

and heres the use of the server in gremlin:

https://gist.github.com/eni9889/d215b2effaac75c2678d

**Comment by newenglander** [comment]
I did notice one difference in the code for Gremlin: `addDownload:` is called before `setDownloadHandler:`. This doesn't affect duplicate entries, but when importing a lot of files in sequence, it increase the success rate (fewer or even no errors about any files that could not be imported).

**Comment by newenglander** [comment]
@eni9889: so you just used a webserver on the iPhone and gave a localhost-based url to `setPrimaryAssetURL:`, or did you have to change anything else? I just tried this out quickly but didn't have any luck.

**Comment by eni9889** [comment]
Turns out that it still doesnt work I don't know what I changed but I started getting the same behavior again but I know that gremlin uses a server bc those logs are from a gremlin import.

**Comment by Tatsh** [comment]
Any luck on this with iOS 6.1.2? I have seen various apps pop up including Bolt and Bridge but none have a feature to import multiple files at once easily (Bridge has it but who wants to tap the screen a thousand times?). Is there possibly a reason for this?

**Comment by newenglander** [comment]
@Tatsh it seems that if you import too many files at once without pausing between (or knowing when one import is done), you get a lot of import failures. What's the Bolt app you mentioned? I can only find a "Statusbar battery icon replacement." (Sorry if we're going off topic here)

**Comment by Tatsh** [comment]
Bolt is here, relatively new, and available on Cydia. Basically it does the exact same thing as Bridge. http://boltapp.org/

I'm interested in furthering the process of getting music synced up in an easy way (and hopefully a way that would make iTunes take those into its library on a computer). The issue has been that none of the apps really do multiple files easily. Bridge does, but then again it has no 'batch mode' like 'load an entire directory'.

I tried reversing Bridge a bit but I'm not sure on my GDB version on iOS. All the real code is in `_mh_execute_header()` of course so it's not easy to dissect there. However `class-dump` works fine but there's nothing extremely useful. I used cycript in an attempt to run some of its functionality manually while it was open but there is still some steps necessary to make this work that I cannot see fully (although I can get to the point in the code and look at the assembly, but it's very hard to decode in such a large block).

These are my notes I gathered on Bridge: https://gist.github.com/Tatsh/f5be6414daab0598b290

**Comment by Tatsh** [comment]
The main thing I would like to be able to do is avoid iTunes altogether for music syncing and completely switch to using Clementine (I have write access to the project). Unlike the libimobiledevice project, I do not particularly care if this requires having a jailbroken device.

**Comment by newenglander** [comment]
@Tatsh Thanks for the link. Send me an email (on my profile page) and I'll give you my two cents, I don't know if continuing the conversation here is welcome. This is after all a bug reporting forum ;)

**Comment by H2CO3** [comment]
This conversation is the most welcome here guys :) Maybe you would find useful my advice: if you try looking at Bridge, hook (using MSHookFunction) the "NSSelectorFromString()" function. That serves as the core of the "obfuscation" its author implements.

**Comment by Tatsh** [comment]
Hah. Did he just use a bunch of C/C++ to avoid `class-dump` reversing?

**Comment by H2CO3** [comment]
Nah, he "disguised" Objective-C method calls by not calling them directly but through [object performSelector:NSSelectorFromString:@"foo"]

**Comment by Tatsh** [comment]
Yeah then one easy hack is to run cycript on the process to get the address of `NSSelectorFromString()` (gdb might find this too since it's a simple C function) and then break on it in gdb. Continue, press import?, and when it stops, `p (char *)$r1` should have the selector.

**Comment by eni9889** [comment]
Any luck on fixing duplicates using the gremlin methods?

**Comment by Tatsh** [comment]
I think Bridge is the thing that can help us the most here for now @eni9889

**Comment by eni9889** [comment]
hmm i didn't even know that existed does Bridge use SSDownload as well?

**Edit seems he does and heres the dump this might be easier to reverse:
https://gist.github.com/eni9889/6b6991df4494da79f32c

**Comment by Tatsh** [comment]
Very nice dump there, but I don't see where it actually adds the 'purchase' to the database. At the very least, perhaps we know now:
- make a `SSDownloadMetadata` object
- pass it to `-[SSDownload initWithDownloadMetadata:(SSDownloadMeta *)metadata]`
- `[SSDownloadManager addDownloads:arrayOfSSDownloadObjects]` ? Does this start the 'downloads' asynchronously?
- `SSDownloadManager` and `SSDownloadQueue` seem to do the rest of the work

**Comment by eni9889** [comment]
I'm pretty sure that addDownloads is what starts the download. The other thing is I think he subclassed SSDownloadManager to JGDownloadManager and he doesn't allow it to download twice. I'm currently working on that part. Also, the line Feb 27 20:57:25 iPhone5Dev SpringBoard[4796] <Warning>: -[<SSDownloadManager: 0x1edca6c0> _handleDownloadsRemoved:0x1eaf28f0] kind of hints to me that one of the downloads is canceled 

I hope we can figure this out b/c Gremlin while it says is open source is really only open source when we figure out what the actual code does on our own which sucks. And honestly this ability to import media into itunes should be available to every app and or tweak without charge and the fact that so many are trying to profit from it is sad.

**Comment by Tatsh** [comment]
@eni9889 any luck on that private method?

**Comment by eni9889** [comment]
I have an exam Monday so I stopped working on it but I did start a project to go along with what I said earlier https://github.com/eni9889/UniversalMediaImporter

**Comment by Krivoblotsky** [comment]
Guys, any update on this issue? I'm having the same issue with duplicates in my project. Did someone overcome this problem?

**Comment by newenglander** [comment]
@Krivoblotsky Wouldn't it be more useful to focus on iOS 7 now? Or does this happen there too? I was thinking something else would have changed so that this no longer works at all.

**Comment by caylorme** [comment]
This bug exists in iOS7 for A7 devices (tested iPhone 5S and iPad Air)

**Comment by newenglander** [comment]
@caylorme For me, this doesn't work on iOS 7 (iPhone 4) at all, using the tweak and app (Music2iPod) in Cydia (as well as similar code in another app). Did you have to change anything to get the import working?

**Comment by caylorme** [comment]
@newenglander try mew seek and grooveshark enhancer. Both import with duplicates. I think Bridge is doing the same.

**Comment by newenglander** [comment]
@caylorme Older versions of Bridge imported duplicates, but the latest version doesn't--tried it on my iPhone 4 running iOS 7. Grooveshark Enhancer uses Gremlin, and this library hasn't been updated for iOS 7 yet (https://github.com/yfrancis/gremlin/issues/6). My apps used it too, and I had no problems with duplicates in iOS 6. Mewseek doesn't seem to be iOS 7 ready yet.
